src/main/java/com/hierynomus/smbj/connection/Connection.java

```diff
@@ -71,21 +72,36 @@
     private static final Logger logger = LoggerFactory.getLogger(Connection.class);
     private ConnectionInfo connectionInfo;
 
+    private SMBClient client;
+    public SMBClient getClient() {
+        return client;
+    }
+
     private Config config;
     private TransportLayer transport;
     private final SMBEventBus bus;
     private PacketReader packetReader;
     private Thread packetReaderThread;
     private final ReentrantLock lock = new ReentrantLock();
 
-    public Connection(Config config, TransportLayer transport, SMBEventBus bus) {
+    public Connection(Config config, SMBClient client, TransportLayer transport, SMBEventBus bus) {
```

Comment:
The SMBClient parameter is there so we can eventually call client.connect() on a new server as part of processing a CREATE message (when we get a PATH_NOT_COVERED error).

The context stack as it is now is like this:  a client owns multiple connections, a connection owns multiple sessions, a session owns multiple treeConnects, and most of the messages that are sent over the transport are sent in this context.
To resolve the PATH_NOT_COVERED error, we have to be able to take this context, but then create a new connection/session/treeConnect/transport owned by the same client.

Another way of thinking about it is that we have a link one direction:
client->connection->session->treeConnect->message
and we need to be able to go backward when the message returns the PATH_NOT_COVERED error:
client<-connection<-session<-treeConnect<-message
because we need to create a new stack:
client->connection2->session2->treeConnect2->message2
to be able to send the message to the new server (and to also send the DFS referral query messages to intermediate servers).